src/main/java/seedu/mark/model/autotag/SelectiveBookmarkTagger.java

```diff
@@ -0,0 +1,49 @@
+package seedu.mark.model.autotag;
+
+import java.util.function.Predicate;
+
+import seedu.mark.model.bookmark.Bookmark;
+import seedu.mark.model.tag.Tag;
+
+/**
+ * A type of BookmarkTagger that tags specific Bookmarks only, based on
+ * whether they match a given {@code Predicate}.
+ */
+public class SelectiveBookmarkTagger extends BookmarkTagger {
+    private final Predicate<Bookmark> predicate;
+
+    public SelectiveBookmarkTagger(Tag tag, Predicate<Bookmark> predicate) {
+        super(tag);
+        this.predicate = predicate;
+    }
+
+    private Predicate<Bookmark> getPredicate() {
+        return this.predicate;
+    }
+
+    /**
+     * Tests the given {@code Bookmark} and tags it only if it matches this
+     * {@code SelectiveBookmarkTagger}'s {@code predicate}.
+     *
+     * @param bookmark A bookmark to test and possibly tag.
+     * @return A new tagged Bookmark if the bookmark matches the predicate, and
+     *         the original Bookmark otherwise.
+     */
+    public Bookmark applyTagSelectively(Bookmark bookmark) {
+        return predicate.test(bookmark) ? super.applyTag(bookmark) : bookmark;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
```

Comment:
Can just put `return super && getPredicate().equals(((SelectiveBookmarkTagger) other).getPredicate())` I think?

Reply:
Actually why do we need the BookmarkTagger? Can we combine BookmarkTagger and SelectiveBookmarkTagger?

Reply:
I think I replied to the comment you deleted :') 

Anyway, I separated BookmarkTagger and SelectiveBookmarkTagger because of their different functions (it seemed a bit strange to have a `BookmarkTagger` that doesn't always tag the bookmarks it's given). 

`BookmarkTagger` is more general, so I was thinking that it might come in useful elsewhere in the code base (not sure exactly where, but it could be used anytime that we need to tag Bookmarks). Of course, if we're moving `BookmarkBuilder` from test to main, we might be able to use that instead. What do you think?

Reply:
Oops! 
I get where you are coming from. Let's keep this first. I think it wont take a lot work to remove it in the future if we want to.